### PR TITLE
Implement a conditional authenticator to check if a sub-flow was executed or not previously in the process

### LINKED ISSUE #35231: Ability to reject authentication to users without 2FA configured
### Description

The customer has a business requirement to take certain actions if users are not configured for 2FA. They are looking to perform an authentication Flow that uses this condition to detect if a user does not have any 2FA method.

### Possible solution

The easiest approach is to implement the option `Negate output` in the condition `Condition - User configured` . The option `Negate output` is available for example in the `Condition - user role` .

Currently the `Condition - User configured` evaluates to true if at least one of the 2-factor authenticators configured in the related conditional subflow is present. So when added `Negate output` option, it will allow to have subflow, which will evaluate to true just if NONE of the configured 2-factor authenticators from the same subflow is present.

Together with the `Deny Access` authenticator, this can be used to achieve the requirement. As for example it may be possible to configure the authentication flow with the subflows like these:

```
- (All other authenticators like cookie, password etc)
- Subflow1 - 2factor enabled
  - Condition - user configured (Negate output = false)
  - OTP ALTERNATIVE
  - WebAuthn ALTERNATIVE
  - Recovery codes ALTERNATIVE
- Subflow2 - 2factor disabled
  - Condition - user configured (Negate output = true)
  - Deny Access ALTERNATIVE
  - OTP ALTERNATIVE
  - WebAuthn ALTERNATIVE
  - Recovery codes ALTERNATIVE 
```

This will allow to authenticate users with 2-factor configured (subflow1), but reject authentication for users without 2-factor configured (subflow2).

It is possible that some other adjustements may be needed in the implementation of `ConditionalUserConfiguredAuthenticator` to make this happen or configure the `subflow2` in slightly different way. But hopefully some setup like this may work. Will need further investigation...

The disadvantage is the need of duplication of all the 2-factor authenticators in both subflows and the complication of such a setup. The better will be to have authentication policies, which would allow better achieve something like this more easily. But that is for the future though...

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by rmartinc**
Closes #35231

The PR adds a new conditional authenticator to check if a previous sub-flow has been executed (or not executed) successfully during the process. This allows to specify more complex flows and execute steps in case a previous one was executed or missed.

I checked and I don't see a list of authenticators in the docs so I added nothing in the server development doc.

Is a note needed in the release notes? To what version? Do we want to add this in 26.0 or just 26.1? (Adding the condition is not a big change but maybe it's better to document this in the release.)

### CODE REVIEW

**Comment by mposolda** [comment]
@rmartinc Thanks, Looks nice in my opinion!

Agree that documentation can be nice. Maybe something to the "Server admin guide" - https://www.keycloak.org/docs/latest/server_admin/index.html#configuring-authentication_server_administration_guide (I can imagine example with multiple 2FA in the 1st conditional subflow and then 2nd subflow with `ResetOTP` authenticator , which means the guy without any 2FA is required to configure OTP during authentication. But different example is completely OK as well, so whatever you prefer) 

Also release notes could be nice as well.

IMO this is for Keycloak 26.1 . It is not bugfix, but rather new thing, so no need to backport this. People just need to wait for new capabilities IMO.

**Review by mposolda — DISMISSED** [review]
@rmartinc Thanks!

There is another conflict in release notes for 26.1.0.

@andymunro Do you please have a chance to review this one please?

**Comment by rmartinc** [comment]
Rebased! Notes are always a hot area! :smile:

File: docs/documentation/release_notes/topics/26_1_0.adoc [hunk 1]
```diff
@@ -51,6 +51,10 @@ by the LDAP provider.
 
 As OpenShift v3 reached end-of-life a while back, support for identity brokering with OpenShift v3 has been removed from Keycloak.
 
+= New conditional authenticator `Condition - sub-flow executed`
+
+The `Condition - sub-flow executed` is a new conditional authenticator in {project_name}. The condition checks if a previous sub-flow was executed (or not executed) successfully during the authentication flow execution. More information about conditions in the link:{adminguide_link}#conditions-in-conditional-flows[server admin guide].
```
Comment (andymunro) [inline docs/documentation/release_notes/topics/26_1_0.adoc:56]:
```suggestion
The `Condition - sub-flow executed` is a new conditional authenticator in {project_name}. The condition checks if a previous sub-flow was executed (or not executed) successfully during the authentication flow execution. For more details, see link:{adminguide_link}#conditions-in-conditional-flows[Conditions in conditional flows].
```

File: docs/documentation/server_admin/topics/authentication/conditions.adoc [hunk 2]
```diff
@@ -45,6 +45,16 @@ Negate output:::
 You can negate the output.
 In other words, the attribute should not be present.
 
+`Condition - sub-flow executed`::
+The condition checks if a previous sub-flow was successfully executed (or not executed) in the authentication process. This way, the flow can trigger some other steps depending on a previous sub-flow termination. Configuration fields:
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/authentication/conditions.adoc:49]:
```suggestion
The condition checks if a previous sub-flow was successfully executed (or not executed) in the authentication process. Therefore, the flow can trigger other steps based on a previous sub-flow termination. Configuration fields:
```

File: docs/documentation/server_admin/topics/authentication/conditions.adoc [hunk 3]
```diff
@@ -85,4 +95,36 @@ The last thing is defining the property with an error message in the login theme
...
+
+==== 2FA conditional workflow examples
+
+The section presents some examples of conditional workflows that integrates 2nd Factor Authentication (2FA) in different ways. The examples copy the default `browser` flow and modify the configuration inside the `forms` sub-flow.
+
+===== Conditional 2FA sub-flow
+
+The default `browser` flow uses a `Conditional OTP` sub-flow that already gives a 2FA with OTP Form (One Time Password). Following the same idea, different 2FA methods can be integrated with the `Condition - User Configured`.
+
+image:images/2fa-example1.png[2FA all alternative]
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/authentication/conditions.adoc:108]:
```suggestion
,2FA all alternative
image:images/2fa-example1.png[2FA all alternative]
```

File: docs/documentation/server_admin/topics/authentication/conditions.adoc [hunk 4]
```diff
@@ -85,4 +95,36 @@ The last thing is defining the property with an error message in the login theme
...
+
+The `forms` sub-flow contains another `2FA` conditional sub-flow with `Condition - user configured`. Three 2FA steps (OTP, Webauthn and Recovery Codes) are allowed as alternative steps. The user will be able to choose one of the three options, if they are configured for the user. As the sub-flow is conditional, the authentication process will complete successfully if no 2FA credential is configured.
+
+===== Conditional 2FA sub-flow and deny access
+
+The second example continues the previous one. After the `2FA` sub-flow, another flow `Deny access if no 2FA` is used to check if the previous `2FA` was not executed. In that case (the user has no 2FA credential configured) the access is denied.
+
+image:images/2fa-example2.png[2FA all alternative and deny access]
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/authentication/conditions.adoc:116]:
```suggestion
.2FA all alternative and deny access
image:images/2fa-example2.png[2FA all alternative and deny access]
```

File: docs/documentation/server_admin/topics/authentication/conditions.adoc [hunk 5]
```diff
@@ -85,4 +95,36 @@ The last thing is defining the property with an error message in the login theme
...
+
+The `Condition - sub-flow executed` is configured to detect if the `2FA` sub-flow was not executed previously.
+
+image:images/2fa-example2-config.png[Configuration for the sub-flow executed]
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/authentication/conditions.adoc:120]:
```suggestion
.Configuration for the sub-flow executed
image:images/2fa-example2-config.png[Configuration for the sub-flow executed]
```

File: docs/documentation/server_admin/topics/authentication/conditions.adoc [hunk 6]
```diff
@@ -85,4 +95,36 @@ The last thing is defining the property with an error message in the login theme
...
+
+The step `Deny access` denies the authentication if not executed.
+
+===== Conditional 2FA sub-flow with OTP default
+
+The last example is very similar to the previous one. Instead of denying the access, step `OTP Form` is configured as required.
+
+image:images/2fa-example3.png[2FA all alternative with OTP default]
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/authentication/conditions.adoc:128]:
```suggestion
.2FA all alternative with OTP default
image:images/2fa-example3.png[2FA all alternative with OTP default]
```

**Review by andymunro — CHANGES_REQUESTED** [review]
@rmartinc A few suggestions. Looks very good.

**Review by mposolda — APPROVED** [review]
@andymunro Thanks for the review!

**Pull request merged**